ompi/mca/coll/libpnbc_osc/pnbc_osc_allreduce.c

```diff
@@ -0,0 +1,1048 @@
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * Author(s): Torsten Hoefler <htor@cs.indiana.edu>
+ *
+ */
+#include "pnbc_osc_internal.h"
+#include "ompi/communicator/communicator.h"
+#include "ompi/datatype/ompi_datatype.h"
+#include "ompi/op/op.h"
+#include "opal/util/bit_ops.h"
+
+#include <assert.h>
+
+static inline int allred_sched_diss(int rank, int p, int count, MPI_Datatype datatype, ptrdiff_t gap,
+                                    const void *sendbuf, void *recvbuf, MPI_Op op, char inplace,
+                                    NBC_Schedule *schedule, void *tmpbuf, ompi_request_t ** request);
+static inline int allred_sched_ring(int rank, int p, int count, MPI_Datatype datatype,
+                                    const void *sendbuf, void *recvbuf, MPI_Op op, int size, int ext,
+                                    NBC_Schedule *schedule, void *tmpbuf);
+static inline int allred_sched_linear(int rank, int p, const void *sendbuf, void *recvbuf, int count,
+                                      MPI_Datatype datatype, ptrdiff_t gap, MPI_Op op, int ext,
+                                      int size, NBC_Schedule *schedule, void *tmpbuf);
+
+static inline int allred_sched_redscat_allgather( int rank, int comm_size, int count,
+                                                  MPI_Datatype datatype, ptrdiff_t gap,
+                                                  const void *sbuf, void *rbuf, MPI_Op op,
+                                                  char inplace, NBC_Schedule *schedule,
+                                                  void *tmpbuf, struct ompi_communicator_t *comm);
+
+
+static int nbc_allreduce_init(const void* sendbuf, void* recvbuf, int count, MPI_Datatype datatype,
+                              MPI_Op op, struct ompi_communicator_t *comm, ompi_request_t ** request,
+                              struct mca_coll_base_module_2_3_0_t *module, bool persistent)
+{
+  int rank, p, res;
+  ptrdiff_t ext, lb;
+  NBC_Schedule *schedule;
+  size_t size;
+
+  enum { NBC_ARED_BINOMIAL, NBC_ARED_RING, NBC_ARED_REDSCAT_ALLGATHER } alg;
+  char inplace;
+  void *tmpbuf = NULL;
+  ompi_coll_libnbc_module_t *libnbc_module = (ompi_coll_libnbc_module_t*) module;
+  ptrdiff_t span, gap;
+
+  NBC_IN_PLACE(sendbuf, recvbuf, inplace);
+
+  rank = ompi_comm_rank (comm);
+  p = ompi_comm_size (comm);
+
+  /* checks for datatype extent */
+  res = ompi_datatype_get_extent(datatype, &lb, &ext);
+  if (OMPI_SUCCESS != res) {
+    NBC_Error ("MPI Error in ompi_datatype_type_extent() (%i)", res);
+    return res;
+  }
+  /* checks for datatype size */
+  res = ompi_datatype_type_size (datatype, &size);
+  if (OMPI_SUCCESS != res) {
+    NBC_Error ("MPI Error in ompi_datatype_type_size() (%i)", res);
+    return res;
+  }
+
+  if (1 == p && (!persistent || inplace)) {
+    if (!inplace) {
+      /* for a single node - copy data to receivebuf */
+      res = NBC_Copy(sendbuf, count, datatype, recvbuf, count, datatype, comm);
```

Comment:
This is incorrect in the NBC code that you are using as a starting point for this new function.

The `copy` must be done again each time this persistent operation is started - it is not a one-off action done during initialisation. This line of code should add a copy step to the schedule, i.e. it should be calling `NBC_Sched_copy` and then committing the schedule.